### PR TITLE
Ticket 266

### PR DESCRIPTION
**Pull request by cfraizer**
- [ ] Updated related documentations
- [ ] Added the change to the Changelog

- I wasn't sure how you wanted to document this, so I didn't.
- The only significant change is in `PHApp`'s `focus` method.

Of course, I'd be happy to make any changes you suggest.

### CODE REVIEW

File: Phoenix/PHApp.m [hunk 1]
```diff
@@ -166,7 +167,27 @@ - (BOOL) activate {
...
+        
+
+        ProcessSerialNumber process;
+        OSStatus error = GetProcessForPID(self.app.processIdentifier, &process);
+        if (error) {
+            return false;
+        }
+
+        error = SetFrontProcessWithOptions(&process, kSetFrontProcessFrontWindowOnly);
+        return (error == 0) ? true : false;
```
Comment (kasper) [inline Phoenix/PHApp.m:189]: You can just return the result of the conditional (i.e. no need for the explicit ternary).

File: Phoenix/PHApp.m [hunk 2]
```diff
@@ -166,7 +167,27 @@ - (BOOL) activate {
 
 - (BOOL) focus {
 
-    return [self.app activateWithOptions:NSApplicationActivateIgnoringOtherApps];
+    if (![NSProcessInfo isOperatingSystemAtLeastBigSur]) {
+
+        return [self.app activateWithOptions:NSApplicationActivateIgnoringOtherApps];
+    } else {
+
+            // Source https://stackoverflow.com/a/65464683/525411
```
Comment (kasper) [inline Phoenix/PHApp.m:175]:
```suggestion
        // FIX: Workaround for the buggy focus behaviour in Big Sur, see issue #266. For reference see https://stackoverflow.com/a/65464683/525411.
```

**Review by kasper — COMMENTED** [review]
Nice, thank you! Added some things to address, could you also add the change to the CHANGELOG?

File: Phoenix.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -748,7 +748,7 @@
 				COMBINE_HIDPI_IMAGES = YES;
 				CURRENT_PROJECT_VERSION = 92;
 				ENABLE_HARDENED_RUNTIME = YES;
-				GCC_TREAT_WARNINGS_AS_ERRORS = YES;
```
[the remark this answers is not in the document]
Comment (cfraizer) [reply Phoenix.xcodeproj/project.pbxproj:751]:
```
phoenix/Phoenix/PHApp.m:179:26: error: 'GetProcessForPID' is deprecated: first deprecated in macOS 10.9
      [-Werror,-Wdeprecated-declarations]
        OSStatus error = GetProcessForPID(self.app.processIdentifier, &process);
```

```
phoenix/Phoenix/PHApp.m:184:16: error: 'SetFrontProcessWithOptions' is deprecated: first deprecated in macOS 10.9
      [-Werror,-Wdeprecated-declarations]
        return SetFrontProcessWithOptions(&process, kSetFrontProcessFrontWindowOnly);
```

[the remark this answers is not in the document]
Comment (cfraizer) [reply Phoenix.xcodeproj/project.pbxproj:751]: Nvm. I used `#pragma` to suppress just those two warnings.

**Comment by cfraizer** [comment]
I incorporated your changes, but I made one additional change.

I found that some windows (those from Emacs.app) would fail in `PHAXUIElement`'s `processIdentifier` method.
So I changed line 174 in `PHApp.m` to call `[self.app processIdentifier]`.

(I'm a total Mac programming n00b, so I apologize if I've made this more difficult than it should be! I really appreciate all your help with this.)

**Comment by cfraizer** [comment]
Hmn. Something is still wrong. I don't really understand it.

Now PHWindow's focused method is broken. That is, `Window.focused()` returns `null`.
My guess is that changing the focus through `SetFrontProcess` doesn't do the right thing that `PHWindow`'s `focused` method looks at.

**Comment by kasper** [comment]
@cfraizer Hmm, could you also verify that it triggers the relevant events for focusing?

**Comment by cfraizer** [comment]
I'm not sure what you mean. (Sorry that I'm a bozo!) It certainly moves the focus—that is, it raises the correct window and that window has keyboard input focus, but after moving it, then `Window.focused()` returns `null`.

**Comment by cfraizer** [comment]
Ah, never mind on that other problem. I was returning the wrong value from `PHApp`'s `focus` method and that uncovered a weird (ancient) bug in my own JavaScript that caused it to try to call the JS `Window.focus()` method on a null object.

I've tested as many variations I can think of and the problem no longer occurs.

**Review by kasper — APPROVED** [review]
Seems reasonable, thanks @cfraizer! I’ll test around before releasing a new version.

**Pull request merged**